webhook/main.go

```diff
@@ -0,0 +1,151 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"os"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	esv1alpha1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/controllers/crds"
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+const (
+	errCreateWebhook = "unable to create webhook"
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = esv1beta1.AddToScheme(scheme)
+	_ = esv1alpha1.AddToScheme(scheme)
+}
+
+func checkCerts(certDir, dnsName string) {
```

Comment:
My lack of knowledge with go concurrency doesn't allow me to go any further than this... Any pointers to improve? :D

Reply:
IMO:
1. it shouldn't `os.Exit()` to make it unit-testable, it can return an error. The caller should decide what to do with it
2. we can do a clean shutdown of the webhook server using two sources: when the cert expires and on SIGINT/TERM.

this is untested:

```go
func main(){
  // ...
  // check certs on startup
  err := checkCerts(...)
  if err != nil {
    ...
  }

  ctx, cancel := context.WithCancel(context.Background())
  go func(){

    // notify on sigterm and every 24h
    sigs := make(chan os.Signal, 1)
    signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
    ticker := time.NewTicker(time.Hour *24)
    for {
      select {
        case <- sigs:
          cancel()
        case <- ticker.C:
          err = checkCert()
          if err != nil {
            cancel()
          }
      }
    }
  }()
  // ...
  mgr.Start(ctx)
}
```